llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -8308,6 +8393,156 @@ llvm::Function *SPIRVToLLVM::createLibraryEntryFunc() {
+    return nullValue;
+  }
+
+  SPIRVStorageClassKind pointerStorageClass = toBeWrapped->getType()->getPointerStorageClass();
+
+  // We don't want bounds checks if we are outside private or function storage class.
+  if (pointerStorageClass != StorageClassPrivate && pointerStorageClass != StorageClassFunction) {
+    return nullValue;
+  }
+
+  // Fetch the access chain behind the pointer the memory operation uses.
+  auto const accessChainBase = static_cast<SPIRVAccessChainBase *const>(toBeWrapped);
+
+  std::vector<SPIRVValue *> indicesVec;
+
+  // Find any integer index, either for accessing the array or a struct member.
+  for (auto *const operand : accessChainBase->getOperands()) {
+    if (operand->getType()->isTypeInt()) {
+      indicesVec.push_back(operand);
+    }
+  }
+
+  if (!indicesVec.empty()) {
+    ArrayRef<SPIRVValue *> indices = indicesVec;
+    SPIRVValue *currentIndex = nullptr;
+
+    auto consumeFirstIndex = [&currentIndex, &indices]() {
+      assert(!indices.empty());
+      currentIndex = indices[0];
+      indices = indices.drop_front();
+    };
+
+    auto baseType = accessChainBase->getBase()->getType()->getPointerElementType();
+    consumeFirstIndex();
+
+    // This will contain our final compare result.
+    Value *finalCmpResult = nullptr;
+
+    auto entryBlock = getBuilder()->GetInsertBlock();
+    Function *parent = entryBlock->getParent();
+    auto intermediateBlock = BasicBlock::Create(*m_context, "", parent);
+    auto finalBlock = BasicBlock::Create(*m_context, "", parent);
+
+    // Build a chain of OOB checks.
+    while (currentIndex) {
+      // Handle either arrays or (nested) structs.
+      if (baseType->getOpCode() == OpTypeArray) {
+        if (const auto baseArray = static_cast<const SPIRVTypeArray *>(baseType)) {
+          // Add the bounds check.
+          auto cmpResult = getBuilder()->CreateICmpULT(
+              transValue(currentIndex, getBuilder()->GetInsertBlock()->getParent(), getBuilder()->GetInsertBlock()),
+              getBuilder()->getInt32(baseArray->getArrayLength()));
+
+          if (finalCmpResult) {
+            finalCmpResult = getBuilder()->CreateAnd(finalCmpResult, cmpResult);
+          } else {
+            finalCmpResult = cmpResult;
+          }
+
+          baseType = baseType->getArrayElementType();
+          auto const newOpCode = baseType->getOpCode();
+
+          // Check if the next element is feasible for checking.
+          if (newOpCode != OpTypeArray && newOpCode != OpTypeStruct) {
+            break;
+          }
+        }
+      } else if (baseType->getOpCode() == OpTypeStruct) {
+        // Fetch the struct member index.
+        assert(isConstantOpCode(currentIndex->getOpCode()));
+
+        uint64_t const structMemberIndex = static_cast<SPIRVConstant *>(currentIndex)->getZExtIntValue();
+        baseType = baseType->getStructMemberType(static_cast<size_t>(structMemberIndex));
```

Comment:
I think this would not work if there was a non-arrayed element inside a structure (e.g. "struct BaseStruct { vec4 d; };"), because there would not be an index to consume in consumeFirstIndex() below.

Reply:
Yes, you are correct. I moved the check at the end of the `OpTypeArray` part to the outer loop and added a test. This should fix this issue instead of generating an assertion failure.

Reply:
Maybe I'm missing something, but it looks it would be more natural to re-flow the loop to avoid that extra check altogether.

while (baseType->getOpCode() == OpTypeArray || baseType->getOpCode() == OpTypeStruct){
  consumeFirstIndex();
  ...
  // rest of the code
}